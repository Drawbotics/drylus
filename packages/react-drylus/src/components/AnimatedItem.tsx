import { Transition, Variant, motion } from 'framer-motion';
import React from 'react';

import { Direction, Speed } from '../enums';
import { Responsive, Style } from '../types';
import { useResponsiveProps } from '../utils';

export const itemVariants = {
  initial: {
    opacity: 0,
  },
  visible: {
    opacity: 1,
  },
  small: {
    scale: 0.7,
  },
  up: {
    y: 20,
  },
  down: {
    y: -20,
  },
  right: {
    x: -20,
  },
  left: {
    x: 20,
  },
  enter: {
    scale: 1,
    opacity: 1,
    y: 0,
    x: 0,
  },
};

export function getVariantFromDirection(direction?: Direction): keyof typeof itemVariants {
  switch (direction) {
    case Direction.TOP_DOWN:
      return 'down';
    case Direction.BOTTOM_UP:
      return 'up';
    case Direction.LEFT_RIGHT:
      return 'right';
    case Direction.RIGHT_LEFT:
      return 'left';
    default:
      return 'small';
  }
}

export function getStaggerFromSpeed(speed?: Speed): number {
  switch (speed) {
    case Speed.SLOW:
      return 0.3;
    case Speed.FAST:
      return 0.1;
    default:
      return 0.2;
  }
}

export function getSettingsFromSpeed(speed?: Speed) {
  switch (speed) {
    case Speed.FAST:
      return {
        damping: 20,
        mass: 0.7,
        stiffness: 250,
      };
    case Speed.SLOW:
      return {
        stiffness: 350,
        damping: 25,
        mass: 2,
      };
    default:
      return {};
  }
}

export interface AnimationVariants {
  initial?: Variant;
  enter?: Variant;
  exit?: Variant;
}

export interface AnimatedItemProps {
  /** The elements that will be animated */
  children: React.ReactNode;

  /**
   * Animates when unmounting, only possible if AnimationGroup wraps the component
   * The exit animation will be the opposite of the enter animation i.e. if scale 0.5 -> 1, then exit = 1 -> 0.5
   * This option can be used to individually animate one item within AnimationGroup, or disabling it as well if the option is set on the AnimationGroup
   */
  animateExit?: boolean;

  /**
   * If no direction is provided, the item uses the "scale up, fade in" animation
   * With direction, the component is moved (in the given direction) and faded in, no scale
   */
  direction?: Direction;

  /**
   * Speed of the animation
   * @default DEFAULT
   */
  speed?: Speed;

  /**
   * Used to override each animation variant, see https://www.framer.com/api/motion/animation/#variants
   * Each variant can be overriden separately: `initial`, `enter` (framer `animate` equivalent), `exit`
   */
  variants?: AnimationVariants;

  /** To override the default transition settings (these also change depending on the speed and direction) */
  transition?: Transition;

  /** To override the root `div` element styles if needed */
  style?: Style;

  /** Reponsive prop overrides */
  responsive?: Responsive<this>;
}

export const AnimatedItem = ({ responsive, ...rest }: AnimatedItemProps) => {
  const {
    children,
    speed = Speed.DEFAULT,
    direction,
    style,
    transition = {},
    variants = {},
  } = useResponsiveProps<AnimatedItemProps>(rest, responsive);

  const { exit: customExit = {}, enter: customEnter = {}, initial: customInitial = {} } = variants;

  const transitionOptions = {
    type: 'spring',
    damping: 17,
    stiffness: 400,
    ...getSettingsFromSpeed(speed),
    ...transition,
  };

  return (
    <motion.div
      style={style}
      initial={['initial', getVariantFromDirection(direction), 'customInitial']}
      animate={['enter', 'customEnter']}
      exit={['exit', 'customExit']}
      variants={{ ...itemVariants, customEnter, customExit, customInitial }}
      transition={transitionOptions}>
      {children}
    </motion.div>
  );
};

export interface AnimationGroupProps {
  /** Should be of AnimatedItem type */
  children:
    | React.ReactElement<typeof AnimatedItem>
    | Array<React.ReactElement<typeof AnimatedItem> | null>
    | React.ReactNode;

  /**
   * Will set a delay between each AnimatedItem child
   * @default true
   */
  staggerChildren: boolean;

  /**
   * If true, any AnimatedItem will animate when unmounting, unless explicitely disabled by setting animateExit=false on the item
   * Note 1: for this to work, the AnimationGroup needs to be OUTSIDE the component being removed from the DOM
   * Note 2: this propagates to ALL AnimatedItems within the Group
   */
  animateExit?: boolean;
}

export const AnimationGroup = ({ children }: AnimationGroupProps) => {
  return <motion.div>{children}</motion.div>;
};
