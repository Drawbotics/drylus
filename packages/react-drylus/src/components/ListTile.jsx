import React from 'react';
import { css, cx } from 'emotion';
import PropTypes from 'prop-types';
import sv from '@drawbotics/style-vars';

import Flex, { FlexItem } from '../layout/Flex';


const styles = {
  root: css`
    display: inline-block;
  `,
  clickable: css`
    &:hover {
      cursor: pointer;
    }
  `,
  leading: css`
    display: flex;
    align-items: center;
    margin-right: ${sv.marginExtraSmall};
  `,
  trailing: css`
    display: flex;
    align-items: center;
  `,
  title: css`
    color: ${sv.colorPrimary};
  `,
  subtitle: css`
    color: ${sv.colorSecondary};
    font-size: 0.9rem;
  `,
  withMargin: css`
    margin-bottom: calc(${sv.marginExtraSmall} / 2);
  `,
};


const ListTile = ({
  title,
  subtitle,
  leading,
  trailing,
  onClick,
}) => {
  return (
    <div className={cx(styles.root, { [styles.clickable]: !! onClick })} onClick={onClick}>
      <Flex>
        {do {
          if (leading) {
            <FlexItem>
              <div className={styles.leading}>
                {leading}
              </div>
            </FlexItem>
          }
        }}
        <FlexItem flex>
          {do {
            if (title != null) {
              <div className={cx(styles.title, { [styles.withMargin]: subtitle })}>
                {title}
              </div>
            }
          }}
          {do {
            if (subtitle != null) {
              <div className={styles.subtitle}>
                {subtitle}
              </div>
            }
          }}
        </FlexItem>
        {do {
          if (trailing != null) {
            <FlexItem>
              <div className={styles.trailing}>
                {trailing}
              </div>
            </FlexItem>
          }
        }}
      </Flex>
    </div>
  );
};


ListTile.propTypes = {
  /** Main text to be displayed */
  title: PropTypes.string,

  /** Smaller text displayed below the title */
  subtitle: PropTypes.string,

  /** Can be anything, will appear centered to the left of the title and/or subtitle */
  leading: PropTypes.node,

  /** Can be anything, will appear centered to the right of the title and/or subtitle */
  trailing: PropTypes.node,

  /** Triggered when the component is clicked */
  onClick: PropTypes.func,
};


export default ListTile;
