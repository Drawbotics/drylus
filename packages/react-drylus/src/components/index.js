export Button, { ButtonTiers } from './Button';
export Icon from './Icon';
export Footer from './Footer';
export Navbar from './Navbar';
export Sidebar from './Sidebar';
export Title from './Title';
export RoundIcon from './RoundIcon';
export Paragraph from './Paragraph';
export SegmentedControl from './SegmentedControl';
export Table, { TCell, TBody, TRow, THead } from './Table';
export Label from './Label';
export Panel, { PanelHeader, PanelBody, PanelFooter, PanelSection } from './Panel';
export Pagination from './Pagination';
export Badge from './Badge';
export Drawer from './Drawer';
