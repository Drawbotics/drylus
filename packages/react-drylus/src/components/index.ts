export * from './Button';
export * from './Icon';
export * from './Title';
export * from './RoundIcon';
export * from './Paragraph';
export * from './SegmentedControl';
export * from './Table';
export * from './Label';
export * from './Panel';
export * from './Pagination';
export * from './Badge';
export * from './Drawer';
export * from './Filter';
export * from './TabNavigation';
export * from './Avatar';
export * from './Text';
export * from './Dot';
export * from './Tooltip';
export * from './Collapsible';
export * from './TextLink';
export * from './Modal';
export * from './Tag';
export * from './Spinner';
export * from './BigRadio';
export * from './BigCheckbox';
export * from './Toggle';
export * from './Separator';
export * from './CircularProgress';
export * from './EmptyState';
export * from './Banner';
export * from './Callout';
export * from './ButtonLink';
export * from './AlertsProvider';
export * from './SplashScreen';
export * from './Breadcrumbs';
export * from './Dropdown';
export * from './List';
export * from './Map';
export * from './FilterGroup';
export * from './Popover';
export * from './Tile';
export * from './ProgressBar';
export * from './LoadingPlaceholder';
export * from './SteppedProgressBar';
export * from './UploadBox';
export * from './AttachmentBox';
export * from './FileIcon';
