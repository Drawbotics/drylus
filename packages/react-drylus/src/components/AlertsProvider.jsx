import sv from '@drawbotics/drylus-style-vars';
import { css, cx } from 'emotion';
import PropTypes from 'prop-types';
import React, { useEffect, useReducer, useState } from 'react';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import v4 from 'uuid/v4';

import { styles as themeStyles } from '../base/ThemeProvider';
import { Category, Size, Tier } from '../enums';
import { Flex, FlexAlign, FlexItem, FlexJustify, Margin } from '../layout';
import { getEnumAsClass } from '../utils';
import Button from './Button';
import Icon from './Icon';

const styles = {
  provider: css`
    position: fixed;
    z-index: 99999;
    bottom: ${sv.defaultMargin};
    left: ${sv.defaultMargin};
    display: inline-flex;
    flex-direction: column-reverse;
    align-items: flex-start;
  `,
  root: css`
    display: inline-block;
    position: relative;
    padding: ${sv.paddingExtraSmall};
    padding-left: ${sv.defaultPadding};
    background: ${sv.white};
    border-radius: ${sv.defaultBorderRadius};
    box-shadow: ${sv.elevation2}, inset 0px 0px 0px 1px ${sv.neutralLight};
    max-width: 400px;
    color: ${sv.colorPrimary};
    overflow: hidden;

    [data-element='text'] {
      font-size: 0.95rem;
    }

    [data-element='icon'] {
      margin-top: -1px;
    }

    &::after {
      content: ' ';
      position: absolute;
      left: 0;
      top: 0;
      width: 5px;
      height: 100%;
    }
  `,
  danger: css`
    [data-element='icon'] {
      color: ${sv.red};
    }

    &::after {
      background: ${sv.red};
    }
  `,
  info: css`
    [data-element='icon'] {
      color: ${sv.blue};
    }

    &::after {
      background: ${sv.blue};
    }
  `,
  warning: css`
    [data-element='icon'] {
      color: ${sv.orange};
    }

    &::after {
      background: ${sv.orange};
    }
  `,
  success: css`
    [data-element='icon'] {
      color: ${sv.green};
    }

    &::after {
      background: ${sv.green};
    }
  `,
  alertEnter: css`
    opacity: 0;
    transform: translateY(-5px);
  `,
  alertEnterActive: css`
    opacity: 1;
    transform: translateY(0);
    transition: all ${sv.defaultTransitionTime} ${sv.bouncyTransitionCurve};
  `,
  alertEnterDone: css``,
  alertExit: css`
    opacity: 1;
    transform: translateY(0);
    transition: all ${sv.defaultTransitionTime} ${sv.bouncyTransitionCurve};
  `,
  alertExitActive: css`
    opacity: 0.01;
    transform: translateY(5px);
    transition: all ${sv.defaultTransitionTime} ${sv.bouncyTransitionCurve};
  `,
};

function _getIconForCategory(category) {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}

export const Alert = ({ id, text, category, onClickDismiss, hideDelay }) => {
  const icon = _getIconForCategory(category);

  useEffect(() => {
    if (category !== Category.DANGER) {
      setTimeout(() => onClickDismiss(id), hideDelay);
    }
  }, []);

  return (
    <div
      className={cx(styles.root, {
        [styles[getEnumAsClass(category)]]: category,
      })}>
      <Flex justify={FlexJustify.START} align={FlexAlign.START}>
        <FlexItem>
          <Margin size={{ right: Size.SMALL, top: Size.EXTRA_SMALL }}>
            <div data-element="icon">
              <Icon name={icon} />
            </div>
          </Margin>
        </FlexItem>
        <FlexItem flex>
          <Margin size={{ top: Size.EXTRA_SMALL, bottom: Size.EXTRA_SMALL }}>
            <span data-element="text">{text}</span>
          </Margin>
        </FlexItem>
        <FlexItem>
          <Margin size={{ left: Size.DEFAULT }}>
            <Button
              size={Size.SMALL}
              onClick={() => onClickDismiss(id)}
              tier={Tier.TERTIARY}
              leading={<Icon name="x" />}
            />
          </Margin>
        </FlexItem>
      </Flex>
    </div>
  );
};

Alert.propTypes = {
  /** Text shown within the alert */
  text: PropTypes.string.isRequired,

  category: PropTypes.oneOf([Category.DANGER, Category.SUCCESS, Category.INFO, Category.WARNING])
    .isRequired,

  /** Triggered when the dismiss button is clicked */
  onClickDismiss: PropTypes.func,

  /** If you need to manually hide the alert, you can pass your own ID to call hide */
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

  /** Amount of milliseconds before the alert is dismissed (except for danger) */
  hideDelay: PropTypes.number,
};

Alert.defaultProps = {
  hideDelay: 4000,
};

const Context = React.createContext();

function reducer(alerts, action) {
  const { type, payload } = action;
  if (type === 'showAlert') {
    return [...alerts, payload.alert];
  } else if (type === 'hideAlert') {
    return alerts.filter((a) => payload.id !== a.id);
  } else {
    return alerts;
  }
}

const AlertsProvider = ({ children }) => {
  const [outletElement, setOutletElement] = useState(null);
  const [alerts, dispatch] = useReducer(reducer, []);

  const hideAlert = (id) => {
    dispatch({ type: 'hideAlert', payload: { id } });
  };

  const showAlert = (alertProps) => {
    const id = v4();
    const alert = { id, ...alertProps };
    dispatch({ type: 'showAlert', payload: { alert } });
  };

  useEffect(() => {
    if (!document.getElementById('alerts-outlet')) {
      const alertsOutlet = document.createElement('div');
      alertsOutlet.id = 'alerts-outlet';
      document.body.appendChild(alertsOutlet);
      setOutletElement(alertsOutlet);
    } else {
      setOutletElement(document.getElementById('alerts-outlet'));
    }

    return () => {
      if (outletElement) {
        document.body.removeChild(outletElement);
      }
    };
  }, []);

  if (!outletElement) return null;

  return (
    <Context.Provider value={{ showAlert, hideAlert }}>
      {children}
      {ReactDOM.createPortal(
        <div className={themeStyles.root}>
          <TransitionGroup className={styles.provider}>
            {alerts.map((alert) => (
              <CSSTransition
                key={alert.id}
                timeout={{
                  enter: 500,
                  exit: 300,
                }}
                classNames={{
                  enter: styles.alertEnter,
                  enterActive: styles.alertEnterActive,
                  exit: styles.alertExit,
                  exitActive: styles.alertExitActive,
                }}>
                <Margin size={{ top: Size.SMALL }}>
                  <Alert onClickDismiss={(id) => hideAlert(id, alerts)} {...alert} />
                </Margin>
              </CSSTransition>
            ))}
          </TransitionGroup>
        </div>,
        document.getElementById('alerts-outlet'),
      )}
    </Context.Provider>
  );
};

AlertsProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default AlertsProvider;

export function useAlert() {
  return React.useContext(Context);
}
