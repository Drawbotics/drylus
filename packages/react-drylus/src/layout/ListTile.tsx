import sv from '@drawbotics/drylus-style-vars';
import { css, cx } from 'emotion';
import React from 'react';

import { Text } from '../components';
import { Responsive, Style } from '../types';
import { run, useResponsiveProps } from '../utils';
import { Flex, FlexItem } from './Flex';

const styles = {
  root: css`
    display: inline-block;
  `,
  clickable: css`
    &:hover {
      cursor: pointer;
    }
  `,
  leading: css`
    display: flex;
    align-items: center;
    margin-right: ${sv.marginExtraSmall};
  `,
  trailing: css`
    display: flex;
    align-items: center;
    margin-left: ${sv.marginExtraSmall};
  `,
  title: css`
    color: ${sv.colorPrimary};
  `,
  subtitle: css`
    color: ${sv.colorSecondary};
    font-size: 0.9rem;
  `,
  withMargin: css`
    margin-bottom: calc(${sv.marginExtraSmall} / 2);
  `,
};

interface ListTileProps {
  /** Main text to be displayed */
  title: React.ReactNode;

  /** Smaller text displayed below the title */
  subtitle?: React.ReactNode;

  /** Can be anything, will appear centered to the left of the title and/or subtitle */
  leading?: React.ReactNode;

  /** Can be anything, will appear centered to the right of the title and/or subtitle */
  trailing?: React.ReactNode;

  /** Triggered when the component is clicked */
  onClick?: () => void;

  /** Used for style overrides */
  style?: Style;

  /** Reponsive prop overrides */
  responsive?: Responsive<this>;
}

export const ListTile = ({ responsive, ...rest }: ListTileProps) => {
  const { title, subtitle, leading, trailing, onClick, style } = useResponsiveProps<ListTileProps>(
    rest,
    responsive,
  );

  return (
    <div
      style={style}
      className={cx(styles.root, { [styles.clickable]: !!onClick })}
      onClick={onClick}>
      <Flex>
        {run(() => {
          if (leading) {
            return (
              <FlexItem>
                <div className={styles.leading}>{leading}</div>
              </FlexItem>
            );
          }
        })}
        <FlexItem flex>
          {run(() => {
            if (title != null) {
              if (typeof title === 'string' || (title as React.ReactElement).type === Text) {
                return (
                  <div className={cx(styles.title, { [styles.withMargin]: subtitle != null })}>
                    {title}
                  </div>
                );
              } else {
                return <div className={styles.withMargin}>{title}</div>;
              }
            }
          })}
          {run(() => {
            if (subtitle != null) {
              if (typeof subtitle === 'string' || (subtitle as React.ReactElement).type === Text) {
                return <div className={styles.subtitle}>{subtitle}</div>;
              } else {
                return subtitle;
              }
            }
          })}
        </FlexItem>
        {run(() => {
          if (trailing != null) {
            return (
              <FlexItem>
                <div className={styles.trailing}>{trailing}</div>
              </FlexItem>
            );
          }
        })}
      </Flex>
    </div>
  );
};
