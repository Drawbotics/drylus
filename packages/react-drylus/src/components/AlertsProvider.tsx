import sv from '@drawbotics/drylus-style-vars';
import { css, cx } from 'emotion';
import React, { useEffect, useReducer, useState } from 'react';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import { v4 } from 'uuid';

import { themeStyles } from '../base/ThemeProvider';
import { Category, Size, Tier } from '../enums';
import { Flex, FlexAlign, FlexItem, FlexJustify, Margin } from '../layout';
import { getEnumAsClass, getIconForCategory } from '../utils';
import { Button } from './Button';
import { Icon } from './Icon';

const styles = {
  provider: css`
    position: fixed;
    z-index: 99999;
    bottom: ${sv.defaultMargin};
    left: ${sv.defaultMargin};
    display: inline-flex;
    flex-direction: column-reverse;
    align-items: flex-start;
  `,
  root: css`
    display: inline-block;
    position: relative;
    padding: ${sv.paddingExtraSmall};
    padding-left: ${sv.defaultPadding};
    background: ${sv.white};
    border-radius: ${sv.defaultBorderRadius};
    box-shadow: ${sv.elevation2}, inset 0px 0px 0px 1px ${sv.neutralLight};
    max-width: 400px;
    color: ${sv.colorPrimary};
    overflow: hidden;

    [data-element='text'] {
      font-size: 0.95rem;
    }

    [data-element='icon'] {
      margin-top: -1px;
    }

    &::after {
      content: ' ';
      position: absolute;
      left: 0;
      top: 0;
      width: 5px;
      height: 100%;
    }
  `,
  danger: css`
    [data-element='icon'] {
      color: ${sv.red};
    }

    &::after {
      background: ${sv.red};
    }
  `,
  info: css`
    [data-element='icon'] {
      color: ${sv.blue};
    }

    &::after {
      background: ${sv.blue};
    }
  `,
  warning: css`
    [data-element='icon'] {
      color: ${sv.orange};
    }

    &::after {
      background: ${sv.orange};
    }
  `,
  success: css`
    [data-element='icon'] {
      color: ${sv.green};
    }

    &::after {
      background: ${sv.green};
    }
  `,
  alertEnter: css`
    opacity: 0;
    transform: translateY(-5px);
  `,
  alertEnterActive: css`
    opacity: 1;
    transform: translateY(0);
    transition: all ${sv.defaultTransitionTime} ${sv.bouncyTransitionCurve};
  `,
  alertEnterDone: css``,
  alertExit: css`
    opacity: 1;
    transform: translateY(0);
    transition: all ${sv.defaultTransitionTime} ${sv.bouncyTransitionCurve};
  `,
  alertExitActive: css`
    opacity: 0.01;
    transform: translateY(5px);
    transition: all ${sv.defaultTransitionTime} ${sv.bouncyTransitionCurve};
  `,
};

export interface AlertProps {
  /** Text shown within the alert */
  text: string;

  /** @kind Category  */
  category: Category.DANGER | Category.SUCCESS | Category.INFO | Category.WARNING | Category.BRAND;

  /** Triggered when the dismiss button is clicked */
  onClickDismiss?: (id?: string) => void;

  /** If you need to manually hide the alert, you can pass your own ID to call hide */
  id?: string;

  /**
   * Amount of milliseconds before the alert is dismissed (except for danger)
   * @default 4000
   */
  hideDelay?: number;
}

export const Alert = ({ id, text, category, onClickDismiss, hideDelay = 4000 }: AlertProps) => {
  const icon = getIconForCategory(category);

  useEffect(() => {
    if (category !== Category.DANGER && onClickDismiss != null) {
      setTimeout(() => onClickDismiss(id), hideDelay);
    }
  }, []);

  return (
    <div
      className={cx(styles.root, {
        [styles[getEnumAsClass<typeof styles>(category)]]: category != null,
      })}>
      <Flex justify={FlexJustify.START} align={FlexAlign.START}>
        <FlexItem>
          <Margin size={{ right: Size.SMALL, top: Size.EXTRA_SMALL }}>
            <div data-element="icon">
              <Icon name={icon} />
            </div>
          </Margin>
        </FlexItem>
        <FlexItem flex>
          <Margin size={{ top: Size.EXTRA_SMALL, bottom: Size.EXTRA_SMALL }}>
            <span data-element="text">{text}</span>
          </Margin>
        </FlexItem>
        <FlexItem>
          <Margin size={{ left: Size.DEFAULT }}>
            <Button
              size={Size.SMALL}
              onClick={onClickDismiss != null ? () => onClickDismiss(id) : undefined}
              tier={Tier.TERTIARY}
              leading={<Icon name="x" />}
            />
          </Margin>
        </FlexItem>
      </Flex>
    </div>
  );
};

export interface AlertsContext {
  showAlert: (props: AlertProps) => void;
  hideAlert: (id?: string) => void;
}

export interface Action {
  type: 'showAlert' | 'hideAlert';
  payload: { alert?: AlertProps; id?: string };
}

const Context = React.createContext<AlertsContext>({} as AlertsContext);

function reducer(alerts: Array<AlertProps>, action: Action): Array<AlertProps> {
  const { type, payload } = action;
  if (type === 'showAlert' && payload.alert != null) {
    return [...alerts, payload.alert];
  } else if (type === 'hideAlert') {
    return alerts.filter((a) => payload.id !== a.id);
  } else {
    return alerts;
  }
}

export interface AlertsProviderProps {
  children: React.ReactNode;
}

export const AlertsProvider = ({ children }: AlertsProviderProps) => {
  const [outletElement, setOutletElement] = useState<HTMLElement>();
  const [alerts, dispatch] = useReducer(reducer, []);

  const hideAlert = (id?: string) => {
    dispatch({ type: 'hideAlert', payload: { id } });
  };

  const showAlert = (alertProps: AlertProps) => {
    const id = v4();
    const alert = { id, ...alertProps };
    dispatch({ type: 'showAlert', payload: { alert } });
  };

  useEffect(() => {
    const outlet = document.getElementById('alerts-outlet');
    if (outlet == null) {
      const alertsOutlet = document.createElement('div');
      alertsOutlet.id = 'alerts-outlet';
      document.body.appendChild(alertsOutlet);
      setOutletElement(alertsOutlet);
    } else {
      setOutletElement(outlet);
    }

    return () => {
      if (outletElement) {
        document.body.removeChild(outletElement);
      }
    };
  }, []);

  if (!outletElement) return null;

  return (
    <Context.Provider value={{ showAlert, hideAlert }}>
      {children}
      {ReactDOM.createPortal(
        <div className={themeStyles.root}>
          <TransitionGroup className={styles.provider}>
            {alerts.map((alert: AlertProps) => (
              <CSSTransition
                key={alert.id}
                timeout={{
                  enter: 500,
                  exit: 300,
                }}
                classNames={{
                  enter: styles.alertEnter,
                  enterActive: styles.alertEnterActive,
                  exit: styles.alertExit,
                  exitActive: styles.alertExitActive,
                }}>
                <Margin size={{ top: Size.SMALL }}>
                  <Alert onClickDismiss={(id) => hideAlert(id)} {...alert} />
                </Margin>
              </CSSTransition>
            ))}
          </TransitionGroup>
        </div>,
        document.getElementById('alerts-outlet') as Element,
      )}
    </Context.Provider>
  );
};

export function useAlert() {
  return React.useContext(Context);
}
