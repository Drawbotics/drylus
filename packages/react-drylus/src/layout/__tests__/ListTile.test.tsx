import React from 'react';
import { create } from 'react-test-renderer';

import { Button } from '../../components';
import { ListTile } from '../ListTile';

describe('ListTile', () => {
  describe('matches snapshot when', () => {
    it('has a title', () => {
      const tree = create(<ListTile title="Title" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a subtitle', () => {
      const tree = create(<ListTile title="Title" subtitle="Subtitle" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a custom title', () => {
      const tree = create(<ListTile title={<Button>Button</Button>} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a trailing and leading component', () => {
      const tree = create(
        <ListTile
          title="Title"
          trailing={<Button>Trailing</Button>}
          leading={<Button>Leading</Button>}
        />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('responds correctly to input and', () => {
    it('triggers when clicked', () => {
      let triggered = false;

      expect(triggered).toBeFalsy();

      const component = create(<ListTile title="Title" onClick={() => (triggered = true)} />);

      component.root.props.onClick();

      expect(triggered).toBeTruthy();
    });
  });
});
