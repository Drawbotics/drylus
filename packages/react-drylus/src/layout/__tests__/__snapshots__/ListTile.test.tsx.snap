// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListTile matches snapshot when has a custom title 1`] = `
.emotion-0 {
  background: #eaeff4;
  color: #172b4e;
  border-radius: 5px;
  padding: calc(8px * 1.5) 24px;
  outline: 0;
  border: 0;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0:hover {
  cursor: pointer;
  background: #d3dce6;
}

.emotion-0:active {
  box-shadow: 0 1px 6px #879ba6 inset;
}

.emotion-0:disabled {
  cursor: not-allowed;
  background: #eaeff4;
  color: rgba(135,155,166,0.5);
  box-shadow: none;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-4 {
  display: inline-block;
}

.emotion-1 {
  margin-bottom: calc(8px / 2);
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="emotion-4"
>
  <div
    className="emotion-3"
  >
    <div
      className="emotion-2"
      style={Object {}}
    >
      <div
        className="emotion-1"
      >
        <button
          className="emotion-0"
        >
          Button
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ListTile matches snapshot when has a subtitle 1`] = `
.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-4 {
  display: inline-block;
}

.emotion-1 {
  color: #879ba6;
  font-size: 0.9rem;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  color: #172b4e;
  margin-bottom: calc(8px / 2);
}

<div
  className="emotion-4"
>
  <div
    className="emotion-3"
  >
    <div
      className="emotion-2"
      style={Object {}}
    >
      <div
        className="emotion-0"
      >
        Title
      </div>
      <div
        className="emotion-1"
      >
        Subtitle
      </div>
    </div>
  </div>
</div>
`;

exports[`ListTile matches snapshot when has a title 1`] = `
.emotion-0 {
  color: #172b4e;
}

.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-3 {
  display: inline-block;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <div
      className="emotion-1"
      style={Object {}}
    >
      <div
        className="emotion-0"
      >
        Title
      </div>
    </div>
  </div>
</div>
`;

exports[`ListTile matches snapshot when has a trailing and leading component 1`] = `
.emotion-0 {
  background: #eaeff4;
  color: #172b4e;
  border-radius: 5px;
  padding: calc(8px * 1.5) 24px;
  outline: 0;
  border: 0;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0:hover {
  cursor: pointer;
  background: #d3dce6;
}

.emotion-0:active {
  box-shadow: 0 1px 6px #879ba6 inset;
}

.emotion-0:disabled {
  cursor: not-allowed;
  background: #eaeff4;
  color: rgba(135,155,166,0.5);
  box-shadow: none;
}

.emotion-3 {
  color: #172b4e;
}

.emotion-4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-9 {
  display: inline-block;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 8px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 8px;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="emotion-9"
>
  <div
    className="emotion-8"
  >
    <div
      className="emotion-2"
      style={Object {}}
    >
      <div
        className="emotion-1"
      >
        <button
          className="emotion-0"
        >
          Leading
        </button>
      </div>
    </div>
    <div
      className="emotion-4"
      style={Object {}}
    >
      <div
        className="emotion-3"
      >
        Title
      </div>
    </div>
    <div
      className="emotion-2"
      style={Object {}}
    >
      <div
        className="emotion-6"
      >
        <button
          className="emotion-0"
        >
          Trailing
        </button>
      </div>
    </div>
  </div>
</div>
`;
